src/HaskellWorks/Data/Network/Ip/Block.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DuplicateRecordFields      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE OverloadedStrings          #-}
+
+module HaskellWorks.Data.Network.Ip.Block
+  ( IpBlock(..)
+  , isValidIpBlock
+  , firstAddress
+  , lastAddress
+  ) where
+
+import Control.Applicative
+import Control.Monad
+import Data.Char
+import Data.Maybe
+import Data.Word
+import GHC.Generics
+import HaskellWorks.Data.Bits.BitWise
+import Text.Read
+
+import qualified Data.Attoparsec.Text                  as AP
+import qualified Data.Bits                             as B
+import qualified Data.Text                             as T
+import qualified HaskellWorks.Data.Network.Ip          as Z
+import qualified HaskellWorks.Data.Network.Ip.Internal as I
+import qualified HaskellWorks.Data.Network.Ip.Ipv4     as I4
+import qualified HaskellWorks.Data.Network.Ip.Ipv6     as I6
+import qualified Text.ParserCombinators.ReadPrec       as RP
+
+newtype IpBlock = IpBlock
+  { block :: Either I4.Ipv4Block I6.Ipv6Block
+  } deriving (Eq, Ord, Generic)
```

Comment:
I think it's better to define this as a sum type rather than a new type.

```
data IpBlock
  = IpBlockV4 Ipv4Block
  | IpBlockV6 Ipv6Block
  deriving (Eq, Ord, Generic)
```